helix-view/src/document.rs

```diff
@@ -586,6 +592,41 @@ impl Document {
         }
     }
 
+    /// Reload the document from its path.
+    pub fn reload(&mut self, view_id: ViewId) -> Result<(), Error> {
+        let encoding = &self.encoding;
+        let path = self.path().filter(|path| path.exists());
+
+        // If there is no path or the path no longer exists.
+        if path.is_none() {
+            return Err(anyhow!("can't find file to reload from"));
+        }
+
+        let mut file = std::fs::File::open(path.unwrap())?;
+        let (mut rope, ..) = from_reader(&mut file, Some(encoding))?;
+        with_line_ending(&mut rope);
+
+        let transaction = helix_core::diff::compare_ropes(self.text(), &rope);
+        self.apply(&transaction, view_id);
+        self.append_changes_to_history(view_id);
```

Comment:
I guess this is one way to do it.  But I was assuming we'd just swap the ropes, and use the transaction only for updating cursor positions and appending to the undo stack.  And then pretty much all other data would be invalidated and recomputed from scratch (e.g. syntax tree, etc.).

But I don't know for sure how easy that is with the current APIs.  If that's too involved, then let's put it on the TODO for someone to tackle after this PR.

@archseer Do you have a good feel for how much of a pain that would be with the current APIs in Helix?

Reply:
Applying a diff is preferable here because it doesn't trash the undo stack I think.

Reply:
Right, I was thinking the transaction would be applied to cursors and the undo stack, so they would both end in the same state as this approach.  Just the `Rope`s would be swapped.

Mostly I want to save time on making the edits to the `Rope`, since applying the diff to the old `Rope` just to make its contents identical to the new `Rope` is kind of redundant--we already have the new `Rope`.

But we can address that later.